src/main/java/com/codeborne/selenide/impl/CollectionElement.java

```diff
@@ -3,30 +3,42 @@
 import com.codeborne.selenide.Condition;
 import com.codeborne.selenide.SelenideElement;
 import com.codeborne.selenide.ex.ElementNotFound;
+import org.openqa.selenium.StaleElementReferenceException;
 import org.openqa.selenium.WebElement;
 
 import java.lang.reflect.Proxy;
+import java.util.List;
 
 import static com.codeborne.selenide.Condition.visible;
 
 public class CollectionElement extends WebElementSource {
-  public static SelenideElement wrap(WebElementsCollection collection, int index) {
+  public static SelenideElement wrap(WebElementsCollection collection, List<WebElement> actualElements, int index) {
     return (SelenideElement) Proxy.newProxyInstance(
         collection.getClass().getClassLoader(), new Class<?>[]{SelenideElement.class},
-        new SelenideElementProxy(new CollectionElement(collection, index)));
+        new SelenideElementProxy(new CollectionElement(collection, actualElements, index)));
   }
 
   private final WebElementsCollection collection;
+  private List<WebElement> actualElements;
   private final int index;
 
-  CollectionElement(WebElementsCollection collection, int index) {
+  CollectionElement(WebElementsCollection collection, List<WebElement> actualElements, int index) {
     this.collection = collection;
+    this.actualElements = actualElements;
     this.index = index;
   }
 
   @Override
   public WebElement getWebElement() {
-    return collection.getActualElements().get(index);
+    try {
+      WebElement el = actualElements.get(index);
+      el.isEnabled(); // check staleness
+
+      return el;
+    } catch (StaleElementReferenceException | IndexOutOfBoundsException e) {
+      actualElements = collection.getActualElements();
+      return actualElements.get(index);
```

Comment:
Sorry maybe I get something wrong.
In https://github.com/codeborne/selenide/pull/653/files#diff-7fb24574c58e442aadf59a781455f466R35 checks element for staleness, but here not. why not? is it correct assumption that isn't happening here again?

Reply:
Even if element becomes stale in **catch** block between `actualElements` re-evaluating and returning the element, the `StaleElementRefferenceException` will be caught by `SelenideElementProxy`. So then proxy will handle retry of calling some method on our element. So I mean that absence of staleness validation is quite okay in **catch**. It just means that collection is still in progress of loading. All timeout and retrying logics is business of `SelenideElementProxy`.

Just to explain why do we need this try-catch block. 
Actual elements in **try** block may become stale because at this code-point we don't know what has passed this cached list to us and we don't know whether it is really actual or not.
But anyway we hope that list is actual -> and we try to get element that we need. Then we check whether it is stale or not. If not -> just return it.
But if element from cached list became stale or the index of the element is out of bounds -> it means that real collection in browser has been refreshed since this collection has been cached and we reach **catch** block because of one of two mentioned exceptions.
So staleness validation definitely should be present in **try** block. Because it is the only place where we can signalize that our collection of actual elements is stale.
So in **catch** block we just re-evaluate collection and try to return element that we get. We don't have to add any additional checks here because if smth is wrong -> `SelenideElementProxy` will do the retry for us.

Whew... I hope my explanation is rather clear 😃

Reply:
OK. I got you. Thank you for detailed explanation. I am not familiar with this code and such try\catch's look strange for me😃